File: BOTREST/src/models/menu.py
# src/models/menu.py
from datetime import datetime
from sqlalchemy import String, Integer, Text, Boolean, Float, DateTime
from sqlalchemy.orm import Mapped, mapped_column
from src.core.database import Base


class Category(Base):
    __tablename__ = "categories"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(255))
    type: Mapped[str] = mapped_column(String(50), default="food")  # food, drink, wine, dessert
    description: Mapped[str] = mapped_column(Text, nullable=True)
    image_url: Mapped[str] = mapped_column(String(500), nullable=True)
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, nullable=True)


class MenuItem(Base):
    __tablename__ = "menu_items"

    id: Mapped[int] = mapped_column(primary_key=True)
    category_id: Mapped[int] = mapped_column(Integer, index=True)
    name: Mapped[str] = mapped_column(String(255))
    description: Mapped[str] = mapped_column(Text, nullable=True)
    price: Mapped[float] = mapped_column(Float)
    weight: Mapped[str] = mapped_column(String(50), nullable=True)
    composition: Mapped[str] = mapped_column(Text, nullable=True)
    allergens: Mapped[str] = mapped_column(Text, nullable=True)
    spice_level: Mapped[int] = mapped_column(Integer, nullable=True, default=0)
    is_available: Mapped[bool] = mapped_column(Boolean, default=True)
    is_hit: Mapped[bool] = mapped_column(Boolean, default=False)
    is_new: Mapped[bool] = mapped_column(Boolean, default=False)
    image_url: Mapped[str] = mapped_column(String(500), nullable=True)
    
    # Винные поля
    vintage: Mapped[str] = mapped_column(String(4), nullable=True)
    region: Mapped[str] = mapped_column(String(100), nullable=True)
    grape_variety: Mapped[str] = mapped_column(String(100), nullable=True)
    wine_type: Mapped[str] = mapped_column(String(50), nullable=True)
    food_pairing: Mapped[str] = mapped_column(Text, nullable=True)
    
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, nullable=True)

File: Web/app/src/services/api/menu.ts
// src/services/api/menu.ts
// Работает с локальным REST API бота (http://localhost:8000/api)

const API_BASE = '/api'

export interface Category {
  id: number
  name: string
  type: 'food' | 'drink' | 'wine' | 'dessert'
  description: string | null
  image_url: string | null
  sort_order: number
  created_at: string
  updated_at: string | null
}

export interface MenuItem {
  id: number
  category_id: number
  name: string
  description: string | null
  price: number
  weight: string | null
  composition: string | null
  allergens: string | null
  spice_level: number | null
  is_available: boolean
  is_hit: boolean
  is_new: boolean
  image_url: string | null
  vintage?: string | null
  region?: string | null
  grape_variety?: string | null
  wine_type?: string | null
  food_pairing?: string | null
  created_at: string
  updated_at: string | null
}

export interface MenuItemWithCategory extends MenuItem {
  category?: Category
}

export const fetchCategories = async (): Promise<Category[]> => {
  const res = await fetch(`${API_BASE}/categories`)
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
  return res.json()
}

export const fetchMenuItems = async (categoryId?: string | number): Promise<MenuItemWithCategory[]> => {
  const params = categoryId ? `?category_id=${categoryId}` : ''
  const res = await fetch(`${API_BASE}/menu-items${params}`)
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
  return res.json()
}

export const createMenuItem = async (item: any): Promise<MenuItem> => {
  const res = await fetch(`${API_BASE}/menu-items`, {
    method: 'POST',
    credentials: 'include',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(item),
  })
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
  return res.json()
}

export const updateMenuItem = async (id: string | number, item: any): Promise<MenuItem> => {
  const res = await fetch(`${API_BASE}/menu-items/${id}`, {
    method: 'PATCH',
    credentials: 'include',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(item),
  })
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
  return res.json()
}

export const deleteMenuItem = async (id: string | number): Promise<void> => {
  const res = await fetch(`${API_BASE}/menu-items/${id}`, { method: 'DELETE', credentials: 'include' })
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
}

export const createCategory = async (category: any): Promise<Category> => {
  const res = await fetch(`${API_BASE}/categories`, {
    method: 'POST',
    credentials: 'include',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(category),
  })
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
  return res.json()
}

export const updateCategory = async (id: string | number, category: any): Promise<Category> => {
  const res = await fetch(`${API_BASE}/categories/${id}`, {
    method: 'PATCH',
    credentials: 'include',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(category),
  })
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
  return res.json()
}

export const deleteCategory = async (id: string | number): Promise<void> => {
  const res = await fetch(`${API_BASE}/categories/${id}`, { method: 'DELETE', credentials: 'include' })
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
}
